### ISSUE TITLE
[tt-wdio] Gracefully Stop Executing with Killed

### ISSUE THREAD

**Issue by emilyrohrbough** [cerner/terra-toolkit#197]
# Bug Report

## Description
Fix tt-wdio runner to honor `control + c` when attemping to kill & stop script execution. Currently it continues to loop the options provided or it does not correctly handle promise rejection when the process is killed via `control + c`. 

## Expected Behavior
`control + c` should stop the script correctly.

**Comment by nramamurth** [comment]
@emilyrohrbough 

Just to be sure, are we trying to avoid the situation shown in the screenshot below here? Also, could you let me know how to reproduce the issue of _'continuing to loop the options provided'_ that the description talks about? 

It looks like the [exitHandler in WDIO launcher](https://github.com/webdriverio/webdriverio/blob/v4.14.1/lib/launcher.js#L460) does take care of terminating the processes that are hanging out.

<img width="760" alt="Screen Shot 2019-04-16 at 2 57 45 PM" src="https://user-images.githubusercontent.com/10090859/56198314-51f65f00-6058-11e9-90d1-7d6be48246f1.png">

**Comment by nramamurth** [comment]
@emilyrohrbough 

Any word on this one?

**Comment by emilyrohrbough** [comment]
If we kill the session mid run might be cleaned up by the wdio launcher, but it can sometimes jump into the second loop of the tt-wdio runner and launch another wdio session. When this happens, the selenium grid has likely not been town down completely and it can cause errors in the Selenium Docker Service. You will not always hit the error. You will have to do some testing to see the exact error seen. But yes, the error above looks like one we should be prevents, the run should exit, not jump to the next test execution.

**Comment by nramamurth** [comment]
@emilyrohrbough 

Great! I was able to reproduce the issue each time when the `suite` is set to `unopinionated`.

<img width="704" alt="Screen Shot 2019-04-23 at 10 15 19 AM" src="https://user-images.githubusercontent.com/10090859/56555308-2ae8e180-65b2-11e9-950f-72f3a78417b3.png">

Let me look further as to why this is happening. Lmk if you can think of something too.

**Comment by nramamurth** [comment]
@emilyrohrbough 

This could in fact potentially happen even for a 'opinionated' run when the users explicitly pass down `formFactors` using the `wdio` command-line runner but I don't think this makes sense for someone to pass down `formFactors` when they already have opinionated tests written, or does it?

```
"tt-wdio-opinionated": "node scripts/wdio/wdio-runner-cli.js --formFactors=['medium','tiny'] --suite opinionated",
```

**Comment by nramamurth** [comment]
I noticed that whenever I have tried and reproduced this issue, the runner has exited with **0**  (_will need to figure out why_ 🕵️‍♂️) and it causes the control to fall into the `else` block and then simply start the next loop - this is evident in the screenshot above.

https://github.com/cerner/terra-toolkit/blob/8e93ad1da66a4253463e079699d3cda077a19f63/scripts/wdio/wdio-runner.js#L31-L43

**Comment by nramamurth** [comment]
**Update:** 

I ran the default unopinionated test suite in `terra-toolkit-js` comprising of [i18n-spec.js](https://github.com/cerner/terra-toolkit/blob/master/tests/wdio/i18n-spec.js) and [theme-spec.js](https://github.com/cerner/terra-toolkit/blob/master/tests/wdio/theme-spec.js) and tried to print out what is happening in the _[wdio launcher](https://github.com/webdriverio/webdriverio/blob/v4.14.1/test/spec/unit/launcher.js)_. I could be wrong with my interpretation but below is my observation so far: 

1. For this particular test suite, for each _locale_ and _form factor_ combination, the wdio launcher creates 2 child processes to run two specs mentioned above and each child process has an [endHandler](https://github.com/webdriverio/webdriverio/blob/v4.14.1/lib/launcher.js#L420) and [exitHandler](https://github.com/webdriverio/webdriverio/blob/v4.14.1/lib/launcher.js#L460) associated with it. 
2. I was able to reproduce the issue every time I tried to abort the session using Ctrl + C at the end 
(or near the end) of execution of the first process. I **assume** that there is some kind of race condition between the `endHandler` and `exitHandler` at this point. 

The original issue is reproducible when `exitHandler` of a child process is hit before `endHandler`. In this case, the other second child process is probably not created yet and the current child process' `endHandler` eventually [resolves with `exitCode` set to 0](https://github.com/webdriverio/webdriverio/blob/v4.14.1/lib/launcher.js#L445) which would cause our wdio runner to continue with the next locale and form factor combination.

**Note:** The console logs 'Processes Count', 'exitCode [BEFORE]', 'exitCode [AFTER]' and 'childProcessExitCode' were all placed inside `endHandler` for screenshots. 

**Exit Handler before End Handler:** ❌

<img width="707" alt="ExitHandler - First" src="https://user-images.githubusercontent.com/10090859/56572483-c8f1a180-65dc-11e9-838b-af63ea9b2a5a.png">


**End Handler before Exit Handler:** ✅

<img width="709" alt="EndHandler - First" src="https://user-images.githubusercontent.com/10090859/56572927-9eecaf00-65dd-11e9-9696-46dd489c91e3.png">

**Abort the session mid-way when the current process has not completed yet:** ✅

<img width="703" alt="Stop Execution - MidWay" src="https://user-images.githubusercontent.com/10090859/56573164-1d495100-65de-11e9-99eb-f80cef0dd964.png">

By now, I am lost in a multitude of promises. 

@emilyrohrbough 

I am not really familiar with _WebdriverIO_ source code,  please let me know if you could think of something that I should possibly look into.

**Comment by emilyrohrbough** [comment]
@nramamurth When I logged this issue, I had assumed this was a bug the tt-wdio script and how we loop locales and form factors. I think wdio is correctly handling the `ctr + c` exit, but I think we are not accounting for it in our looping. The wdio launcher will exit correctly, but the tt-wdio script will attempt to exectute the next iteration in the loop and start a new wdio launcher.

Here are the lines I would suspect are causing the issues:
- https://github.com/cerner/terra-toolkit/blob/master/scripts/wdio/wdio-runner.js#L12
- https://github.com/cerner/terra-toolkit/blob/master/scripts/wdio/wdio-runner.js#L16

**Comment by nramamurth** [comment]
> The wdio launcher will exit correctly, but the tt-wdio script will attempt to execute the next iteration in the loop and start a new wdio launcher.

I'd think it is an expected behaviour because wdio exits with a 0 (just like a regular success scenario) in this situation as well. The `tt-wdio` script has **no** control over knowing the difference between the two cases below imo - please correct me if I am wrong.

- successful run (_wdio_ exits with 0).
- when the run is attempted to be aborted with `ctrl + c` that further leads to the situation in discussion (_wdio_ exits with 0).

**Comment by emilyrohrbough** [comment]
@nramamurth that's interesting. I didn't realize the same exit code was returned even if the run was aborted. Do you have an suggestions on how we can maintain the current tt-wdio functionality but prevent a new test run from launching if the tt-wdio run was aborted with `ctrl + c`?

**Comment by nramamurth** [comment]
@emilyrohrbough

I confirmed that it is the same exit code that is returned and it is evident in the test run output too but I couldn't think of a way how the two scenarios could be differentiated from the test runner perspective. I tried looking if _WebdriverIO_ provides more information to the runner but my limited knowledge ain't helping. 

I was thinking if we could also detect a `SIGINT` when `ctrl + c` is pressed and exit immediately out of the first loop but I haven't really gotten to it and not sure if it is possible because from what I understand, the _Launcher_ instance would have started executing the run by then and I believe that [it can detect the signal](https://github.com/webdriverio/webdriverio/blob/v4.14.1/lib/launcher.js#L162) but not our _Runner_ script per se?

**Comment by emilyrohrbough** [comment]
Yeah, I wonder if we could check for
```
process.on('SIGINT', function () {
  process.exit(somecode); 
});
```
ourselves to determine if we should break the loop

**Comment by emilyrohrbough** [comment]
@nramamurth any updates on this?

**Comment by nramamurth** [comment]
@emilyrohrbough 

Nope, I have to find some time to get back to this.

**Comment by ShettyAkarsh** [comment]
@emilyrohrbough 

Myself tried your suggestion mentioned above to have the process exit code in terra-toolkit when ctrl+c is entered. below is the code.

process.on('SIGINT', () => process.exit(0));

so this works fine when the wdio launcher has executed and the execution is in  terra-toolkit wdio-runner-js. during this time if ctrl+c is entered then the process exits as shown in the below screenshot.

![image](https://user-images.githubusercontent.com/45097016/60033574-727d0e00-96c6-11e9-868d-d5598c627c07.png)

if the ctrl+c is entered while generating the screenshots mainly when the execution is in the webdriverIO launcher then the process exits, but with an error 'Error [ERR_IPC_CHANNEL_CLOSED]: Channel closed'.

![image](https://user-images.githubusercontent.com/45097016/60033825-efa88300-96c6-11e9-8f24-9857ea7c28a4.png)

This is mainly because of we are trying to kill the forked child process in the webdriveIO. while its parent process was already killed.

Please share your thoughts

**Comment by emilyrohrbough** [comment]
@ShettyAkarsh Do you have a link to a branch or some code to go with these errors? It's hard to follow what you have actually tried from this output

**Comment by ShettyAkarsh** [comment]
Here goes the Branch https://github.com/cerner/terra-toolkit/commit/e06dcfabf4d9649f4fa72d88f146458a6ed9e94f

**Comment by ryanthemanuel** [comment]
What about something like this https://github.com/cerner/terra-toolkit/commit/59707cec43aa9352b2a3c61b6f1776a5542804f3?  Sorry to use your branch @ShettyAkarsh.  Writing the code was easier than describing it :) I haven't given it a test yet.  Didn't have time at the end of the day :( Might be a good starting point at least.

**Comment by ShettyAkarsh** [comment]
@ryanthemanuel myself tested the above branch code but was getting the error shown in the below screenshot. this error originating from webdriverIO as even they have the logic to terminate the process for ctrl+c entry [webdriverIO](https://github.com/webdriverio/webdriverio/blob/v4.14.1/lib/launcher.js#L162)

![image](https://user-images.githubusercontent.com/45097016/60188207-8bafc700-984c-11e9-8d8a-9ad6cabaece0.png)

so i have updated the new code in the same [branch](https://github.com/cerner/terra-toolkit/commit/f1ada6e1e496d001500260301a26ab18a14341c7). which seems to be working. for b oth cases when the execution is in the webdriverIO code and in the terra-toolkit code .  let me know your thoughts

**Comment by ryanthemanuel** [comment]
That looks good to me if it works in the scenarios we're concerned about.  @emilyrohrbough?

**Comment by emilyrohrbough** [comment]
This looks great!
